tensorflow/python/kernel_tests/xent_op_test_base.py

```diff
@@ -0,0 +1,362 @@
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import nn_ops
+# The following import is required to register the gradient function.
+from tensorflow.python.ops.nn_grad import _SoftmaxCrossEntropyWithLogitsGrad  # pylint: disable=unused-import
+from tensorflow.python.platform import test
+
+
+class XentOpTestBase(test.TestCase):
+
+  def _opDeterminismEnabled(self):
+    deterministic_ops = os.getenv('TF_DETERMINISTIC_OPS', '0')
+    return deterministic_ops == '1' or deterministic_ops == 'true'
+
+  def _coreOp(self, features, labels):
+    """ Run the core op under test.
+
+    Override this method in a child class for cases (e.g. determinism) where the
+    core op is no longer being used.
+    """
+    return gen_nn_ops.softmax_cross_entropy_with_logits(
+        features=features, labels=labels)
+
+  def _gpuNotAvailable(self):
+    return len(config.list_physical_devices('GPU')) == 0
+
+  def _npXent(self, features, labels, dim=-1):
+    if dim == -1:
+      dim = len(features.shape) - 1
+    one_only_on_dim = list(features.shape)
+    one_only_on_dim[dim] = 1
+    e = np.exp(
+        features - np.reshape(np.amax(features, axis=dim), one_only_on_dim))
+    probs = e / np.reshape(np.sum(e, axis=dim), one_only_on_dim)
+    bp = (probs - labels)
+    l = -np.sum(labels * np.log(probs + 1.0e-20), axis=dim)
+    return l, bp
+
+  # TODO(b/123860949): The values are constant folded for XLA, so placeholders
+  # are needed.
+  def _testXent(self,
+                np_features,
+                np_labels,
+                use_gpu=False,
+                with_placeholders=False,
+                test_backprop=True):
+    np_loss, np_backprop = self._npXent(np_features, np_labels)
+    with self.cached_session(use_gpu=use_gpu) as sess:
+      if with_placeholders:
+        features_placeholder = array_ops.placeholder(np_features.dtype)
+        labels_placeholder = array_ops.placeholder(np_labels.dtype)
+        loss, backprop = self._coreOp(
+            labels=labels_placeholder, features=features_placeholder)
+        tf_loss, tf_backprop = sess.run([loss, backprop],
+                                        feed_dict={
+                                            labels_placeholder: np_labels,
+                                            features_placeholder: np_features
+                                        })
+      else:
+        loss, backprop = self._coreOp(np_features, np_labels)
+        tf_loss, tf_backprop = self.evaluate([loss, backprop])
+    self.assertAllCloseAccordingToType(np_loss, tf_loss, half_rtol=1e-2)
+    if test_backprop:
+      self.assertAllCloseAccordingToType(np_backprop, tf_backprop)
+
+  def _testXentWrapper(self, np_features, np_labels, dim=-1, use_gpu=False):
+    np_loss, _ = self._npXent(np_features, np_labels, dim=dim)
+    with self.cached_session(), test_util.device(use_gpu):
+      # Even in eager mode, the above line will be able to pin ops to CPU.
```

Comment:
What does this mean? What is wrong with passing use_gpu to `cached_session`, as was done before?

Reply:
Maybe I'm mistaken, but from my experience of writing, running, and checking tests (and from looking at the test_util code), if you pass `use_gpu=False` into `self.session()` or `self.cached_session()` and run it on a GPU in eager mode, the ops will not be pinned to the CPU. This is because `self.session()` and `self.cached_session()` are only relevant in graph mode. When we write test code that can be run in either graph or eager mode, there needs to be both a `self.session()` or `self.cached_session()` context and a separate, nested `test_util.device()` context.

The existing code is currently only running in eager mode, and, the way it's written, it will be run twice on GPU if there is a GPU present. The change in the PR makes it run on both CPU and GPU if there is a GPU present, but does not make it less compatible with graph mode (in case we want to run this test in graph mode).

Reply:
The next commit will remove the `self.cached_session()` and the comment on the next line. None of that is needed.